### ISSUE TITLE
Filter + Search together throws type error or resets the search text

### ISSUE THREAD

**Issue by shotor** [AllenFang/react-bootstrap-table#457]
I created my own toolbar, containing a switch and a searchbar.

The switch calls `this.refs.table.handleFilterData` while the searchbar makes use of `this.refs.table.handleSearch`

The problems arise during 2 scenario where you filter first, then search or when you search first then filter.

I've included my work arounds, and am wondering anyone had encountered these issues before and what the recommended way is to fix these issues before starting on a PR.

1) Filter then search will throw:

```TableDataStore.js?0c9e:527 Uncaught TypeError: Cannot read property 'filter' of undefined```

Which is the line:
```
const source = this.filterObj !== null ? this.filter(this.filterObj) : this.data;
```

I can work around this by using the filtered data if a filter object is present
But I have to make sure I set the filerObj back to null in my component when I want to reset the filter
```
var source = _this4.filterObj !== null ? _this4.filteredData : _this4.data;
```

2) Search then filter will ignore the search, this is due to the filter always using the complete data set and not the filtered data set:

TableDataStore:264
```
this.filteredData = data.filter(function (row) {
```

I can work around it by using the filtered data if searchText is set.
Again I have to make sure I set the searchText to '' or null when I clear it.
```
var data = this.searchText ? this.filteredData : this.data;
this.filteredData = data.filter(function (
```
edit: This workaround doesn't solve the scenario:
- Sucesfull search
- Filter that returns 0 results
- Turn off filter (should return previous succesfull search, but won't )

I will have another look tomorrow

**Comment by AllenFang** [comment]
@shotor, thanks your feedback, I'll check this out :)

**Comment by bpnetguy** [comment]
I just ran across the same issue.

**Comment by AllenFang** [comment]
Fix on ```v2.3.3```.
